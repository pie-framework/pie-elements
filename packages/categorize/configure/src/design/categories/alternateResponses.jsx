import React from 'react';
import PropTypes from 'prop-types';
import { withStyles } from '@material-ui/core/styles';
import classNames from 'classnames';
import Category from './category';
import { moveChoiceToAlternate, removeChoiceFromAlternate } from '@pie-lib/pie-toolbox/categorize';
import { RowLabel } from './RowLabel';
import { getMaxCategoryChoices } from '../../utils';

const styles = (theme) => ({
  categories: {
    marginBottom: theme.spacing.unit * 2.5,
  },
  categoriesHolder: {
    display: 'grid',
    gridRowGap: `${theme.spacing.unit}px`,
    gridColumnGap: `${theme.spacing.unit}px`,
  },
  row: {
    display: 'grid',
    gridTemplateColumns: 'repeat(2, 1fr)',
    gridColumnGap: `${theme.spacing.unit}px`,
    alignItems: 'baseline',
    width: '100%',
    marginTop: theme.spacing.unit,
    marginBottom: 2 * theme.spacing.unit,
  },
  rowLabel: {
    gridColumn: '1/3',
  },
  rowLabelHolder: {
    width: '100%',
  },
});

export class AlternateResponses extends React.Component {
  static propTypes = {
    altIndex: PropTypes.number.isRequired,
    configuration: PropTypes.object,
    imageSupport: PropTypes.shape({
      add: PropTypes.func.isRequired,
      delete: PropTypes.func.isRequired,
    }),
    classes: PropTypes.object.isRequired,
    className: PropTypes.string,
    categories: PropTypes.array,
    defaultImageMaxHeight: PropTypes.number,
    defaultImageMaxWidth: PropTypes.number,
    onModelChanged: PropTypes.func,
    model: PropTypes.object.isRequired,
    uploadSoundSupport: PropTypes.shape({
      add: PropTypes.func.isRequired,
      delete: PropTypes.func.isRequired,
    }),
    toolbarOpts: PropTypes.object,
    spellCheck: PropTypes.bool,
  };

  addChoiceToCategory = (addedChoice, categoryId) => {
    const {
      altIndex,
      model: { correctResponse, choices, maxChoicesPerCategory = 0 },
      onModelChanged,
    } = this.props;

    const choice = choices.find((c) => c.id === addedChoice.id);

    correctResponse.forEach((a) => {
      if (a.category === categoryId) {
        a.alternateResponses = a.alternateResponses || [];

        if (!a.alternateResponses[altIndex]) {
          a.alternateResponses[altIndex] = [];
        }

        a.alternateResponses[altIndex].push(addedChoice.id);
        if (choice.categoryCount && choice.categoryCount !== 0) {
          a.alternateResponses[altIndex] = a.alternateResponses[altIndex].reduce((acc, currentValue) => {
            if (currentValue === choice.id) {
              const foundIndex = acc.findIndex((c) => c === choice.id);
              if (foundIndex === -1) {
                acc.push(currentValue);
              }
            } else {
              acc.push(currentValue);
            }

            return acc;
          }, []);
        }

        return a;
      } else {
        if (a.alternateResponses[altIndex] && choice.categoryCount !== 0) {
          a.alternateResponses[altIndex] = a.alternateResponses[altIndex].filter((c) => c !== addedChoice.id);
          return a;
        }
      }

      return a;
    });

    const maxCategoryChoices = getMaxCategoryChoices(this.props.model);
    // when maxChoicesPerCategory is set to 0, there is no limit so it should not be updated
    onModelChanged({
      correctResponse,
      maxChoicesPerCategory:
        maxChoicesPerCategory !== 0 && maxChoicesPerCategory < maxCategoryChoices
          ? maxChoicesPerCategory + 1
          : maxChoicesPerCategory,
    });
  };

  moveChoice = (choiceId, from, to, choiceIndex, alternateIndex) => {
    const { model, onModelChanged } = this.props;
    let { choices, correctResponse = [], maxChoicesPerCategory = 0 } = model || {};
    const choice = (choices || []).find((choice) => choice.id === choiceId);
    correctResponse = moveChoiceToAlternate(
      choiceId,
      from,
      to,
      choiceIndex,
      correctResponse,
      alternateIndex,
      choice?.categoryCount,
    );

    const maxCategoryChoices = getMaxCategoryChoices(this.props.model);
    // when maxChoicesPerCategory is set to 0, there is no limit so it should not be updated
    onModelChanged({
      correctResponse,
      maxChoicesPerCategory:
        maxChoicesPerCategory !== 0 && maxChoicesPerCategory < maxCategoryChoices
          ? maxChoicesPerCategory + 1
          : maxChoicesPerCategory,
    });
  };

  deleteChoiceFromCategory = (category, choice, choiceIndex) => {
    const { model, altIndex, onModelChanged } = this.props;

    const correctResponse = removeChoiceFromAlternate(
      choice.id,
      category.id,
      choiceIndex,
      altIndex,
      model.correctResponse,
    );

    onModelChanged({ correctResponse });
  };

  render() {
    const {
      altIndex,
      model,
      configuration,
      classes,
      className,
      categories,
      imageSupport,
      spellCheck,
      uploadSoundSupport,
      toolbarOpts,
      defaultImageMaxHeight,
      defaultImageMaxWidth,
      mathMlOptions = {},
    } = this.props;
    const { categoriesPerRow, errors, rowLabels } = model;
    const { duplicateAlternate } = errors || {};
    const { maxImageWidth = {}, maxImageHeight = {} } = configuration || {};

    const holderStyle = {
      gridTemplateColumns: `repeat(${categoriesPerRow}, 1fr)`,
    };
    const isDuplicated = duplicateAlternate ? duplicateAlternate.index === altIndex : false;

    return (
      <div className={classNames(classes.categories, className)}>
        <div className={classes.categoriesHolder} style={holderStyle}>
          {categories.map((category, index) => {
            const hasRowLabel = index % categoriesPerRow === 0;
            const rowIndex = index / categoriesPerRow;

            return (
              <React.Fragment key={index}>
                {hasRowLabel && (
                  <div
                    style={{
                      gridColumn: `1/${categoriesPerRow + 1}`,
                      width: '100%',
                    }}
                    className={classes.rowLabel}
                    dangerouslySetInnerHTML={{
                      __html: rowLabels[rowIndex] || '',
                    }}
                  ></div>
                )}

                <Category
                  key={index}
                  alternateResponseIndex={altIndex}
                  imageSupport={imageSupport}
                  isDuplicated={isDuplicated && duplicateAlternate.category === category.id}
                  category={category}
                  spellCheck={spellCheck}
                  onAddChoice={this.addChoiceToCategory}
                  onDeleteChoice={(choice, choiceIndex) => this.deleteChoiceFromCategory(category, choice, choiceIndex)}
                  onMoveChoice={(choiceId, from, to, choiceIndex, alternateIndex) =>
                    this.moveChoice(choiceId, from, to, choiceIndex, alternateIndex)
                  }
                  uploadSoundSupport={uploadSoundSupport}
                  mathMlOptions={mathMlOptions}
                  configuration={configuration}
                />
              </React.Fragment>
            );
          })}
        </div>
      </div>
    );
  }
}

export default withStyles(styles)(AlternateResponses);
