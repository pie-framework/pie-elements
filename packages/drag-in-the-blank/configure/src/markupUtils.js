import escape from 'lodash/escape';
import isUndefined from 'lodash/isUndefined';

const replaceHtmlRegex = /<[^>]*>?/gm;

const createElementFromHTML = htmlString => {
  const div = document.createElement('div');

  div.innerHTML = htmlString.trim();

  return div;
};

export const processMarkup = markup => {
  const newMarkup = markup.replace(/(\\t)|(\\n)/g, '').replace(/\\"/g, '"').replace(/\\\//g, '/');
  const slateMarkup = createElementFromHTML(newMarkup);
  const choices = [];
  let index = 0;

  slateMarkup.querySelectorAll('[data-type="drag_in_the_blank"]').forEach(s => {
    let value = s.dataset.value && s.dataset.value.replace(/&nbsp;/g, ' ').trim();

    if (!value) {
      value = '';
    }

    choices.push({
      value,
      id: s.dataset.id
    });

    s.replaceWith(`{{${index++}}}`);
  });

  return {
    markup: slateMarkup.innerHTML,
    choices: choices,
    correctResponse: choices.reduce((obj, c, index) => {
      obj[index] = !isUndefined(c.id) && c.id || '';

      return obj;
    }, {})
  };
};

const REGEX = /\{\{(\d+)\}\}/g;

export const createSlateMarkup = (markup, choices, correctResponse) => {
  const newMarkup = markup.replace(/(\\t)|(\\n)/g, '').replace(/\\"/g, '"').replace(/\\\//g, '/');
  let index = 0;

  return newMarkup.replace(REGEX, (match, g) => {
    const correctId = correctResponse[g];
    let correctChoice = choices.find(c => c.id === correctId);

    if (!correctChoice || !correctChoice.value) {
      correctChoice = {
        id: '',
        value: ''
      };
    }

    return `<span data-type="drag_in_the_blank" data-index="${index++}" data-id="${correctChoice.id}" data-value="${escape(correctChoice.value)}"></span>`;
  });
};

export const choiceIsEmpty = choice => {
  if (choice) {
    const { value = '' } = choice;

    return value.trim() === '' || value.replace(replaceHtmlRegex, '') === '';
  }

  return false;
};
