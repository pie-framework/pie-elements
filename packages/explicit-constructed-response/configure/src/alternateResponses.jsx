import React from 'react';
import PropTypes from 'prop-types';
import isEqual from 'lodash/isEqual';
import map from 'lodash/map';
import reduce from 'lodash/reduce';

import AlternateSection from './alternateSection';

export class AlternateResponses extends React.Component {
  static propTypes = {
    model: PropTypes.object.isRequired,
    onChange: PropTypes.func.isRequired
  };

  state = {};

  UNSAFE_componentWillReceiveProps(nextProps) {
    this.updateChoicesIfNeeded(nextProps);
  }

  componentDidMount() {
    this.updateChoicesIfNeeded(this.props);
  }

  updateChoicesIfNeeded = props => {
    if (!this.state.choices
      || !isEqual(this.state.choices, props.model.choices)
      || !isEqual(props.model.choices, this.props.model.choices)
    ) {
      const { choices } = props.model;

      const selectedValues = reduce(choices, (obj, c, key) => {
        if (c.length > 1) {
          obj[key] = c[0];
        }

        return obj;
      }, {});

      this.setState({
        choices: props.model.choices,
        values: selectedValues
      });
    }
  };

  getRemainingChoices = valueKey => {
    const { choices } = this.state;
    const result = reduce(choices, (arr, c, key) => {
      if (c.length === 1 && !valueKey) {
        arr.push({
          label: c[0].label,
          value: key
        });
      }

      return arr;
    }, []);

    console.log(result);

    return result;
  };

  onChoiceChanged = (choice, key) => {
    const { onChange } = this.props;
    const { choices } = this.state;
    const sectionChoices = (choices[key] || []);

    const isNew = !sectionChoices.find(c => c.value === choice.value);

    const newChoices = sectionChoices.reduce((arr, c) => {
      const newVal = c.value === choice.value ? choice : c;

      arr.push(newVal);

      return arr;
    }, []);

    if (isNew) {
      newChoices.push(choice);
    }

    onChange({
      ...choices,
      [key]: newChoices
    });
  };

  onChoiceRemoved = (value, section) => {
    const { onChange } = this.props;
    const { choices } = this.state;
    const sectionChoices = (choices[section] || []);

    const newChoices = sectionChoices.reduce((arr, c) => {
      if (c.value === value) {
        return arr;
      }

      arr.push(c);

      return arr;
    }, []);

    onChange({
      ...choices,
      [section]: newChoices
    });
  };

  onSectionSelect = (choice, key) => {
    const { onChange } = this.props;
    const { choices, values } = this.state;

    if (choices[key] && choices[key].length > 1) {
      if (!choice) {
        onChange({
          ...choices,
          [key]: [choices[key][0]]
        });
      }
    } else {
      this.setState({
        choices: {
          ...choices,
          [key]: [
            ...choices[key],
            {
              label: '',
              value: '1'
            }
          ]
        },
        values: {
          ...values,
          [key]: choices[key][0]
        }
      });
    }
  };

  render() {
    const { choices } = this.state;

    return (
      <div>
        {map(choices, (c, key) => {
          if (c.length > 1) {
            const selected = this.state.values[key];

            return (
              <AlternateSection
                key={key}
                value={selected && selected.value}
                onSelect={choice => this.onSectionSelect(choice, key)}
                choiceChanged={choice => this.onChoiceChanged(choice, key)}
                choiceRemoved={value => this.onChoiceRemoved(value, key)}
                selectChoices={[selected]}
                choices={c}
              />
            );
          }
        })}
        {
          choices &&
          Object.keys(this.state.values).length !== Object.keys(choices).length &&
          <AlternateSection
            value=""
            onSelect={choice => this.onSectionSelect(choice, choice.value)}
            choiceChanged={choice => this.onChoiceChanged(choice)}
            choiceRemoved={value => this.onChoiceRemoved(value)}
            selectChoices={this.getRemainingChoices()}
          />
        }
      </div>
    );
  }
}

export default AlternateResponses;
