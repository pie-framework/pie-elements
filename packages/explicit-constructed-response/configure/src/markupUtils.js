import escape from 'lodash/escape';

export const removeUnwantedCharacters = markup =>
  markup
    .replace(/(\t+(?!imes))|(\n)|(\\t+(?!imes))|(\\n)/g, '')
    .replace(/\\"/g, '"').replace(/\\\//g, '/');

const createElementFromHTML = (htmlString = '') => {
  const div = document.createElement('div');

  div.innerHTML = htmlString.trim();

  return div;
};

export const processMarkup = markup => {
  const newMarkup = removeUnwantedCharacters(markup || '');
  const slateMarkup = createElementFromHTML(newMarkup || '');
  let index = 0;

  slateMarkup.querySelectorAll('[data-type="explicit_constructed_response"]').forEach(s => {
    s.replaceWith(`{{${index++}}}`);
  });

  return slateMarkup.innerHTML;
};

const REGEX = /\{\{(\d+)\}\}/g;

export const createSlateMarkup = (markup, choices) => {
  if (!markup) {
    return '';
  }

  const newMarkup = removeUnwantedCharacters(markup);

  return newMarkup.replace(REGEX, (match, g) => {
    const label = choices[g][0].label || '';

    return `<span data-type="explicit_constructed_response" data-index="${g}" data-value="${escape(label)}"></span>`;
  });
};

// also used in controller/src/index.js
export const getAdjustedLength = length => {
  if (length <= 2) {
    return length + 2;
  }

  if (length <= 4) {
    return length + 3;
  }

  if (length <= 6) {
    return length + 4;
  }

  return length + 5;
};
