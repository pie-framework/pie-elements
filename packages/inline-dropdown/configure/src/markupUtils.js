import escape from 'lodash/escape';

const createElementFromHTML = htmlString => {
  const div = document.createElement('div');

  div.innerHTML = htmlString.trim();

  return div;
};

export const processMarkup = markup => {
  const newMarkup = markup.replace(/(\\t)|(\\n)/g, '').replace(/\\"/g, '"').replace(/\\\//g, '/');
  const slateMarkup = createElementFromHTML(newMarkup);
  let index = 0;

  slateMarkup.querySelectorAll('[data-type="inline_dropdown"]').forEach(s => {
    s.replaceWith(` {{${index++}}} `);
  });

  return slateMarkup.innerHTML;
};

const REGEX = /\{\{(\d+)\}\}/g;

/*export const createSlateMarkup = (markup, choices) => {
  const newMarkup = markup.replace(/(\\t)|(\\n)/g, '').replace(/\\"/g, '"').replace(/\\\//g, '/');
  const createMenuItem = (id, val) => `<span data-type="menu_item" data-id="${id}">${val}</span>`;
  const createSelect = choices => {
    let correctChoice = choices.find(c => c.correct);
    const menuItems = choices.map(c => createMenuItem(c.value, c.label));

    if (!correctChoice || !correctChoice.value) {
      correctChoice = {
        id: '',
        value: ''
      };
    }

    return `<span data-type="inline_dropdown" data-correct-id="${correctChoice.value}">${menuItems.join('')}</span>`;
  };

  return newMarkup.replace(REGEX, (match, g) => {
    return createSelect(choices[g]);
  });
};*/
export const createSlateMarkup = (markup, choices) => {
  const newMarkup = markup.replace(/(\\t)|(\\n)/g, '').replace(/\\"/g, '"').replace(/\\\//g, '/');
  const createMenuItem = (id, val) => `<span data-type="menu_item" data-id="${id}">${val}</span>`;
  const createSelect = index => {
    let correctChoice = choices[index].find(c => c.correct);

    if (!correctChoice || !correctChoice.value) {
      correctChoice = {
        id: '',
        value: '',
        label: ''
      };
    }

    return `<span data-type="inline_dropdown" data-index="${index}" data-value="${escape(correctChoice.label)}"></span>`;
  };

  return newMarkup.replace(REGEX, (match, g) => {
    return createSelect(g);
  });
};
