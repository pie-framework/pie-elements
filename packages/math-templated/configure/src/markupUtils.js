// do not remove \t from \times, \triangle, \tan, \theta or \therefore
const tSymbols = 'imes|riangle|an|heta|herefore';
// do not remove \n from \nthroot, \nparallel, \ncong, \napprox, \neq, \ne or \nsim
const nSymbols = 'throot|parallel|cong|approx|eq|e|sim';
// match all \t and \n that are not part of math symbols that starts with \t or \n
const matchTabAndNewLine = new RegExp(
  `(\\t(?!${tSymbols}))|(\\n(?!${nSymbols}))|(\\\\t(?!${tSymbols}))|(\\\\n(?!${nSymbols}))`,
  'g',
);

export const removeUnwantedCharacters = (markup) =>
  markup.replace(matchTabAndNewLine, '').replace(/\\"/g, '"').replace(/\\\//g, '/');

const createElementFromHTML = (htmlString = '') => {
  const div = document.createElement('div');

  div.innerHTML = htmlString.trim();

  return div;
};

export const processMarkup = (markup) => {
  const newMarkup = removeUnwantedCharacters(markup || '');
  const slateMarkup = createElementFromHTML(newMarkup || '');
  let index = 0;

  slateMarkup.querySelectorAll('[data-type="explicit_constructed_response"]').forEach((s) => {
    s.replaceWith(`{{${index++}}}`);
  });

  return slateMarkup.innerHTML;
};

const REGEX = /\{\{(\d+)\}\}/g;

export const createSlateMarkup = (markup) => {
  if (!markup) {
    return '';
  }

  const newMarkup = removeUnwantedCharacters(markup);
  return newMarkup.replace(REGEX, (match, g) =>
    `<span data-type="explicit_constructed_response" data-index="${g}" data-value="response ${parseInt(g) + 1}"></span>`
  );
};
