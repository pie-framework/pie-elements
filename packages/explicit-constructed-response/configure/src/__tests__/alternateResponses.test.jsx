import { shallow } from 'enzyme';
import React from 'react';

import { AlternateResponses } from '../alternateResponses';
import sensibleDefaults from '../defaults';
import { createSlateMarkup, processMarkup } from '../markupUtils';

const model = {
  markup: '<p>The {{0}} jumped {{1}} the {{2}}</p>',
  disabled: false,
  choices: {
    0: [{ label: 'cow', value: '0' }, { label: 'cattle', value: '1' }, { label: 'calf', value: '2' }],
    1: [{ label: 'over', value: '0' }, { label: 'past', value: '1' }, { label: 'beyond', value: '2' }],
    2: [{ label: 'moon', value: '0' }]
  },
  maxLengthPerChoice: [6, 6, 4],
  prompt: 'Complete the sentence',
};

const prepareModel = (model = {}) => {
  const joinedObj = {
    ...sensibleDefaults.model,
    ...model,
  };
  const slateMarkup = joinedObj.slateMarkup || createSlateMarkup(joinedObj.markup, joinedObj.choices);
  const processedMarkup = processMarkup(slateMarkup);

  return {
    ...joinedObj,
    slateMarkup,
    markup: processedMarkup
  };
};

describe('AlternateResponses', () => {
  let onChange = jest.fn();

  const wrapper = extras => {
    const defaults = {
      onChange,
      classes: {},
      model: prepareModel({
        ...model,
        ...extras
      })
    };
    const props = { ...defaults };

    return shallow(<AlternateResponses {...props} />);
  };

  describe('snapshot', () => {
    it('renders', () => {
      expect(wrapper()).toMatchSnapshot();
    });
  });

  describe('logic', () => {
    let w;

    beforeEach(() => {
      w = wrapper();
    });

    describe('updateChoicesIfNeeded', () => {
      it('sets state, updates edited choice', () => {
        const newChoices = {
          0: [{ label: 'cow', value: '0' }, { label: 'cattle', value: '1' }, { label: 'calf', value: '2' }],
          1: [{ label: 'under', value: '0' }, { label: 'past', value: '1' }, { label: 'beyond', value: '2' }],
          2: [{ label: 'moon', value: '0' }]
        };

        w.instance().updateChoicesIfNeeded({
          model: {
            ...model,
            choices: newChoices,
          }
        });

        expect(w.instance().state).toEqual(expect.objectContaining({
          choices: newChoices,
          values: {
            0: { label: 'cow', value: '0' },
            1: { label: 'under', value: '0' }
          }
        }));
      });

      it('sets state, if a choice has an alternate', () => {
        const newChoices = {
          0: [{ label: 'cow', value: '0' }, { label: 'cattle', value: '1' }, { label: 'calf', value: '2' }],
          1: [{ label: 'under', value: '0' }, { label: 'past', value: '1' }, { label: 'beyond', value: '2' }],
          2: [{ label: 'sun', value: '0' }, { label: 'star', value: '1' }]
        };

        w.instance().updateChoicesIfNeeded({
          model: {
            ...model,
            choices: newChoices,
          }
        });

        expect(w.instance().state).toEqual(expect.objectContaining({
          choices: newChoices,
          values: {
            0: { label: 'cow', value: '0' },
            1: { label: 'under', value: '0' },
            2: { label: 'sun', value: '0' }
          }
        }));
      });

      it('sets state if some choices are removed', () => {
        const newChoices = {
          0: [{ label: 'cow', value: '0' }, { label: 'cattle', value: '1' }, { label: 'calf', value: '2' }],
        };

        w.instance().updateChoicesIfNeeded({
          model: {
            ...model,
            choices: newChoices,
          }
        });

        expect(w.instance().state).toEqual(expect.objectContaining({
          choices: newChoices,
          values: {
            0: { label: 'cow', value: '0' }
          }
        }));
      });

      it('does not change state', () => {
        w.instance().updateChoicesIfNeeded({ model });

        expect(w.instance().state).toEqual(expect.objectContaining({
          choices: model.choices,
          values: {
            0: { label: 'cow', value: '0' },
            1: { label: 'over', value: '0' }
          }
        }));
      });
    });

    describe('getRemainingChoices', () => {
      it('adds an alternate if selecting a choice', () => {
        const remainingChoices = w.instance().getRemainingChoices();

        expect(remainingChoices).toEqual([
          { label: 'moon', value: '2' }
        ]);
      });
    });

    describe('onChoiceChanged', () => {
      it('changes new alternate', () => {
        w.instance().onChoiceChanged({ label: 'New Choice', value: '1' }, 2);

        expect(onChange).toBeCalledWith({
            ...model.choices,
            2: [{ label: 'moon', value: '0' }, { label: 'New Choice', value: '1' }]
          }
        );
      });

      it('changes existing alternate', () => {
        w.instance().onChoiceChanged({ label: 'New Choice Edited', value: '1' }, 2);

        expect(onChange).toBeCalledWith({
            ...model.choices,
            2: [{ label: 'moon', value: '0' }, { label: 'New Choice Edited', value: '1' }]
          }
        );
      });
    });

    describe('onChoiceRemoved', () => {
      it('removes choice from section', () => {
        const choices = w.instance().props.model.choices;

        w.instance().onChoiceRemoved('1', '1');

        choices[1] = [{ label: 'over', value: '0' }, { label: 'beyond', value: '2' }];

        expect(onChange).toBeCalledWith(choices);
      });

      it('does not throw error if parameters are not correct', () => {
        const choices = w.instance().props.model.choices;

        w.instance().onChoiceRemoved('11', '1');

        expect(onChange).toBeCalledWith(choices);
      });
    });

    describe('onSectionSelect', () => {
      it('removing a selection', () => {
        const choices = w.instance().props.model.choices;

        w.instance().onSectionSelect(undefined, '0');

        expect(onChange).toBeCalledWith({
          ...choices,
          '0': [{ label: 'cow', value: '0' }]
        });
      });

      it('selecting a response', () => {
        const state = w.instance().state;

        w.instance().onSectionSelect({ label: 'moon', value: '0' }, '2');

        expect(w.instance().state.choices).toEqual({
          ...state.choices,
          '2': [{ label: 'moon', value: '0' }, { label: '', value: '1' }]
        });
        expect(w.instance().state.values).toEqual({
          ...state.values,
          2: { label: 'moon', value: '0' }
        });
      });
    });
  });
});
