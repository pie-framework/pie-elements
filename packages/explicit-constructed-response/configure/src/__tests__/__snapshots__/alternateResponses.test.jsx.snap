// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AlternateResponses snapshot renders 1`] = `
<div>
  <WithStyles(AlternateSection)
    choiceChanged={[Function]}
    choiceRemoved={[Function]}
    choices={
      Array [
        Object {
          "label": "cow",
          "value": "0",
        },
        Object {
          "label": "cattle",
          "value": "1",
        },
        Object {
          "label": "calf",
          "value": "2",
        },
      ]
    }
    key="0"
    lengthChanged={[Function]}
    maxLength={6}
    onSelect={[Function]}
    selectChoices={
      Array [
        Object {
          "label": "cow",
          "value": "0",
        },
      ]
    }
    showMaxLength={true}
    value="0"
  />
  <WithStyles(AlternateSection)
    choiceChanged={[Function]}
    choiceRemoved={[Function]}
    choices={
      Array [
        Object {
          "label": "over",
          "value": "0",
        },
        Object {
          "label": "past",
          "value": "1",
        },
        Object {
          "label": "beyond",
          "value": "2",
        },
      ]
    }
    key="1"
    lengthChanged={[Function]}
    maxLength={6}
    onSelect={[Function]}
    selectChoices={
      Array [
        Object {
          "label": "over",
          "value": "0",
        },
      ]
    }
    showMaxLength={true}
    value="0"
  />
  <WithStyles(AlternateSection)
    choiceChanged={[Function]}
    choiceRemoved={[Function]}
    choices={
      Array [
        Object {
          "label": "moon",
          "value": "0",
        },
      ]
    }
    key="2"
    lengthChanged={[Function]}
    maxLength={4}
    onSelect={[Function]}
    selectChoices={
      Array [
        undefined,
      ]
    }
    showMaxLength={true}
  />
  <WithStyles(AlternateSection)
    choiceChanged={[Function]}
    choiceRemoved={[Function]}
    onSelect={[Function]}
    selectChoices={
      Array [
        Object {
          "label": "moon",
          "value": "2",
        },
      ]
    }
    value=""
  />
</div>
`;
